vortex-array/src/compute/compare.rs

```diff
@@ -194,8 +195,21 @@ fn arrow_compare(
     operator: Operator,
 ) -> VortexResult<ArrayRef> {
     let nullable = left.dtype().is_nullable() || right.dtype().is_nullable();
-    let lhs = Datum::try_new(left.to_array())?;
-    let rhs = Datum::try_new(right.to_array())?;
+    // Make sure both of the arrays end up with the same arrow data type
+    let lhs = if matches!(left.dtype(), DType::Utf8(_)) {
+        Datum::with_target_datatype(left.to_array(), &DataType::Utf8View)?
+    } else if matches!(left.dtype(), DType::Binary(_)) {
+        Datum::with_target_datatype(left.to_array(), &DataType::BinaryView)?
+    } else {
+        Datum::try_new(left.to_array())?
+    };
+    let rhs = if matches!(right.dtype(), DType::Utf8(_)) {
+        Datum::with_target_datatype(right.to_array(), &DataType::Utf8View)?
+    } else if matches!(right.dtype(), DType::Binary(_)) {
+        Datum::with_target_datatype(right.to_array(), &DataType::BinaryView)?
+    } else {
+        Datum::try_new(right.to_array())?
+    };
```

Comment:
It would be tempting to just change Datum::try_new to go via canonical encoding but in some places we use it to convert to arrow without going through canonical :/